### PR TITLE
added bitcoind client & syncing

### PR DESCRIPTION
**Pull request by BatiGencho**
This MR changes the mempool impl to bitcoind introducing a new client and integration for syncing state. There are also clippy fixes and other rusty improvements.

Tickets:

- https://github.com/botanix-labs/botanix/issues/342
- https://github.com/botanix-labs/botanix/issues/343

### CODE REVIEW

File: crates/consensus/authority/src/block_builder.rs [hunk 1]
```diff
@@ -34,6 +34,19 @@ where
         // Check if we are in_turn
         let is_inturn = self.epoch_manager.poll().await;
 
+        match self.bitcoind_client.is_synced().await {
```
Comment (arminsabouri) [inline crates/consensus/authority/src/block_builder.rs]: This code looks correct but im not sure whether it belong here or in the main function. i.e is there any point to starting reth until the bitcoind client is sync'd? Maybe we should busy loop until this is the case in main. I'll leave this as a open question. Lmk what you think

**Review by arminsabouri — COMMENTED** [review]
I just had one open question. Awesome work!

File: crates/consensus/authority/src/block_builder.rs (hunk 1 above)
Comment (BatiGencho) [reply crates/consensus/authority/src/block_builder.rs]: I though of it too, actually wanted to leave a note. It definitely makes sense to await the sync in the start (poa_builder.rs) - I will move the code there, not an issue. Thanks for pointing this out to me.

Comment (arminsabouri) [reply crates/consensus/authority/src/block_builder.rs]: Great! after that happy to get this merged

Comment (scottmillner) [reply crates/consensus/authority/src/block_builder.rs]: awesome sauce! looks great! will ✅ once sync check is in main function
The Mighty Mutex strikes again!!

**Comment by BatiGencho** [comment]
@0xBEEFCAF3 I have addressed your request. Do you mind going over it again.

File: Makefile [hunk 2]
```diff
@@ -418,7 +418,9 @@ start-poa-server-1:
 	--authrpc.addr "127.0.0.1" \
 	--authrpc.port 8551 \
 	--btc-server "localhost:8080" \
-	--btc-block-source "https://mempool.space/signet/api" \
+	--bitcoind.url "https://bitcoind.botanixlabs.dev" \
+	--bitcoind.username "mempool" \
+	--bitcoind.password "mempool" \
```
Comment (arminsabouri) [inline Makefile]: These are the actual creds for the signet rpc right? Can we read these from env var?

File: crates/btc-wallet/src/bitcoind.rs [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    TransactionBroadcastFailed(bitcoincore_rpc::Error),
+    #[error("Block index failed")]
+    BlockIndexStatusFailed(bitcoincore_rpc::Error),
+}
+
+#[derive(PartialEq, Eq, Debug, Clone, Serialize, Deserialize)]
+pub struct BitcoindConfig {
+    url: Url,
+    username: Option<String>,
+    password: Option<String>,
```
Comment (arminsabouri) [inline crates/btc-wallet/src/bitcoind.rs]: Do these have to be optional? Seems like they are required via the expect() below

File: crates/node-core/src/args/bitcoind_args.rs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+    /// Btcd username
+    ///
+    /// The username of the bitcoind server.
+    #[arg(
+        long = "bitcoind.username",
+        name = "bitcoind.username",
+        value_name = "BITCOIND_USERNAME"
+    )]
+    pub username: Option<String>,
```
Comment (arminsabouri) [inline crates/node-core/src/args/bitcoind_args.rs]: Same thing here. Do the creds have to be optional?

**Review by arminsabouri — CHANGES_REQUESTED** [review]
Everything looks good. Just want to make sure we don't expose any real credentials

File: crates/node-core/src/args/bitcoind_args.rs (hunk 4 above)
Comment (BatiGencho) [reply crates/node-core/src/args/bitcoind_args.rs]: I wasnt sure if bitcoind always needs creds, but that seems to be the case. I have removed the Options now.

File: crates/btc-wallet/src/bitcoind.rs (hunk 3 above)
Comment (BatiGencho) [reply crates/btc-wallet/src/bitcoind.rs]: I wasnt sure if bitcoind always needs creds, but that seems to be the case. I have removed the Options now.

File: Makefile (hunk 2 above)
Comment (BatiGencho) [reply Makefile]: They were, I removed them. Was using them for testing. The vals are now being read from .env file.

**Review by arminsabouri — APPROVED** [review]

**Review by scottmillner — APPROVED** [review]

**Pull request merged**